docs/dev/workflow_report.md

```diff
@@ -0,0 +1,149 @@
+# Workflow report
+
+## Introduction
+Every installer command is internally based on the [assets framework][../design/assetgeneration.md], which allows to define a directed acyclic graph of assets (a generic work item).
+A workflow identifies a set of asset graph paths executed when running a specific command (note that the same asset may behave differently when triggered under different workflows).
+
+A report allows to collect the main relevant workflow events - as well as any eventual error, and detailed results - in a human-readable format, providing an abstraction from the basic asset logging system.
+A report also serializes immediately on the disk any update received, thus providing a simple mechanism to support streaming the progression of the current command execution (particulary
+useful in case of remote execution).
+
+## Stages
+A report is a composed by a number of sequential _stage_. A stage represents a specific workflow phase (given by one or more assets) that could be relevant and/or informative from the user point of view.
+A stage is composed by a _stage identifier_ (or briefly, a stage id) and a number of substages:
+
+* The stage id is defined by a short internal identifier and a longer human-readable description, meant to be shown to the user.
+* Substages are a sequential of zero or more stages, and they are meant to provide a further level of details for the owner stage, if required.
+
+### Stages ids
+A stage id can be defined via the `NewStageID` method, for example:
+```
+StageFetchBaseISO wr.StageID = wr.NewStageID("fetch-base-iso", "Retrieving the base ISO image")
+```
+
+A substage id can be defined by prefixing the internal id with the owner stage id, in the format `<owner stage>.<substage>`, for example:
+```
+StageFetchBaseISOExtract wr.StageID = wr.NewStageID("fetch-base-iso.extract-image", "Extracting base image from release payload")
+StageFetchBaseISOVerify wr.StageID = wr.NewStageID("fetch-base-iso.verify-version", "Verifying base image version")
+StageFetchBaseISODownload wr.StageID = wr.NewStageID("fetch-base-iso.download-image", "Downloading base ISO image")
+```
+
+### Substages
+Substages are completely optional, and they may be useful for detailing a particularly big or lengthy stage. So, it's perfectly fine to define stages without any substage.
+The current framework does not support more than two levels of stages, essentially for keeping both a simpler interface and for producing an easy to read output for the
+final user.
+
+### (Sub)Stage result
```

Comment:
Thanks for adding lots of examples in this doc. An example of how to add a result may be useful here too.

Reply:
Curious, the current implementation doesn't appear to add any results. Are there more changes to come or is it not currently used by the create command?

Reply:
Yes, the ClusterInfo result will be added in another pr to address the separate card https://issues.redhat.com/browse/AGENT-1000